packages/@lwc/engine/src/framework/hot-swaps.ts

```diff
@@ -0,0 +1,226 @@
+    if (process.env.NODE_ENV === 'production') {
+        // this method should never leak to prod
+        throw new ReferenceError();
+    }
+    const visited: Set<StylesheetFactory> = new Set();
+    while (swappedStyleMap.has(style) && !visited.has(style)) {
+        visited.add(style);
+        style = swappedStyleMap.get(style)!;
+    }
+    return style;
+}
+
+export function setActiveVM(vm: VM) {
+    if (process.env.NODE_ENV === 'production') {
+        // this method should never leak to prod
+        throw new ReferenceError();
+    }
+    // tracking active component
+    const Ctor = vm.def.ctor;
+    let componentVMs = activeComponents.get(Ctor);
+    if (isUndefined(componentVMs)) {
+        componentVMs = new Set();
+        activeComponents.set(Ctor, componentVMs);
+    }
+    // this will allow us to keep track of the hot components
+    componentVMs.add(vm);
+
+    // tracking active template
+    const tpl = vm.cmpTemplate;
+    let templateVMs = activeTemplates.get(tpl);
+    if (isUndefined(templateVMs)) {
+        templateVMs = new Set();
+        activeTemplates.set(tpl, templateVMs);
+    }
+    // this will allow us to keep track of the templates that are
+    // being used by a hot component
+    templateVMs.add(vm);
+
+    // tracking active styles associated to template
+    const stylesheets = tpl.stylesheets;
+    if (!isUndefined(stylesheets)) {
+        stylesheets.forEach((stylesheet) => {
+            let stylesheetVMs = activeStyles.get(stylesheet);
+            if (isUndefined(stylesheetVMs)) {
+                stylesheetVMs = new Set();
+                activeStyles.set(stylesheet, stylesheetVMs);
+            }
+            // this will allow us to keep track of the stylesheet that are
+            // being used by a hot component
+            stylesheetVMs.add(vm);
+        });
+    }
+}
+
+export function removeActiveVM(vm: VM) {
+    if (process.env.NODE_ENV === 'production') {
+        // this method should never leak to prod
+        throw new ReferenceError();
+    }
+    // tracking inactive component
+    const Ctor = vm.def.ctor;
+    let list = activeComponents.get(Ctor);
+    if (!isUndefined(list)) {
+        // deleting the vm from the set to avoid leaking memory
+        list.delete(vm);
+    }
+    // removing inactive template
+    const tpl = vm.cmpTemplate;
+    list = activeTemplates.get(tpl);
+    if (!isUndefined(list)) {
+        // deleting the vm from the set to avoid leaking memory
+        list.delete(vm);
+    }
+    // removing active styles associated to template
+    const styles = tpl.stylesheets;
+    if (!isUndefined(styles)) {
+        styles.forEach((style) => {
+            list = activeStyles.get(style);
+            if (!isUndefined(list)) {
+                // deleting the vm from the set to avoid leaking memory
+                list.delete(vm);
+            }
+        });
+    }
+}
+
+export function swapTemplate(oldTpl: Template, newTpl: Template): boolean {
```

Comment:
@diervo @pmdartus @ravijayaramappa do you think we should check the arguments to be a valid template?

Reply:
IMO, no. It will be checked before evaluation in `evaluateTemplate`.
> !isTemplateRegistered(html)
That check is lazy, the template is validated before it is used. So if you register a template and the engine never uses it, the check is avoided.